### DISCUSSION TITLE
Intent to contribute: block malicious install scripts

### DISCUSSION THREAD

**Discussion by naugtur** [vercel/next.js#41022] [Ideas]
### Describe the feature you'd like to request

I'd like to contribute a change to create-next-app to block all install scripts and use https://www.npmjs.com/package/@lavamoat/allow-scripts to securely run the ones that are actually needed for anything.

As a result adding a malicious postinstall script to any of the dependencies that don't already use one would not compromise the app.

### Describe the solution you'd like

I've reviewed the code and it's a matter of modifying the package.json and running an additional step before the first installation in this area somewhere https://github.com/vercel/next.js/blob/canary/packages/create-next-app/create-app.ts#L208

I'm offering to contribute the whole thing.

### Describe alternatives you've considered

lavamoat can be added afterwards by a security conscious developer, but attacks leveraging postinstall have become widespread enough to consider switching to opt-in

**Comment by naugtur** [comment]
As soon as I get confirmation this is potentially interesting, I'll start working on the PR. 
I'd love any and all suggestions or preferences around where and how to implement a change to package.json generation.

**Comment by SukkaW** [comment]
@naugtur 

We need to be very careful since the proposed change is security related.

E.g., it seems that the `@lavamoat/allow-scripts` package has less than 5000 weekly downloads on npm, has it been battle-tested? Has it been reviewed and audited? Is the maintainer of that package trustworthy? Is it possible that that package itself could introduce malicious scripts?

**Reply by Jack-Works** [reply]
Note: `pnpm` has a built-in implementation for this. See [pnpm.onlyBuiltDependencies](https://pnpm.io/package_json#pnpmonlybuiltdependencies).

**Reply by naugtur** [reply]
@SukkaW LavaMoat is a set of tools that is responsible for supply chain security of MetaMask and allow-scripts is just a tiny utility that's part of a much larger toolbox. It's being worked on and released by the security team at MetaMask, myself included.

**Reply by naugtur** [reply]
I'm now close to releasing a solution for https://socket.dev/blog/npm-bin-script-confusion as part of allow-scripts too.

**Comment by sh4nnongoh** [comment]
Would love to see any SES implementation built into the Nextjs framework. Sharing a great introductory video of a SES implementation by Lavamoat below, for those who are unfamiliar with what SES is trying to achieve.

https://youtu.be/Z5Bz0DYga1k

**Reply by SukkaW** [reply]
> Would love to see any SES implementation built into the Next.js framework.

IMHO, I don't see why Next.js should have a built-in SES implementation.
As for the others who need SES, I'd like to recommend https://github.com/DimensionDev/aot-secure-ecmascript (created and maintained by @Jack-Works, a TC39 member and also one of a champion of `proposal-compartments`).

**Reply by Jack-Works** [reply]
If the environment allows `eval`, I would suggest Agorics's [ses](https://npmjs.com/ses) package. My implementation requires

1. the environment has been lockdown by ses (which is very incompatible with the current ecosystem, e.g. `immer`) 
2. a build time compiler, although I see Next.JS already has a build step `next build` so I guess this is OK

**Reply by naugtur** [reply]
My initial suggestion was to only introduce the (totally SES independent) malicious lifecycle scripts protections. I'd not recommend introducing the SES Compartment based protections into a default app template as the dependency tree is large enough to have a decent likelyhood of containing a package that'd have trouble working under lockdown (a form of frozen intrinsics)
